File: tests/rustdoc-js/generics-impl.js
// exact-check

const EXPECTED = [
    {
        'query': 'Aaaaaaa -> u32',
        'others': [
            { 'path': 'generics_impl::Aaaaaaa', 'name': 'bbbbbbb' },
        ],
    },
    {
        'query': 'Aaaaaaa -> bool',
        'others': [
            { 'path': 'generics_impl::Aaaaaaa', 'name': 'ccccccc' },
        ],
    },
    {
        'query': 'Aaaaaaa -> Result<usize>',
        'others': [
            { 'path': 'generics_impl::Aaaaaaa', 'name': 'read' },
        ],
    },
    {
        'query': 'Read -> u64',
        'others': [
            { 'path': 'generics_impl::Ddddddd', 'name': 'eeeeeee' },
        ],
    },
    {
        'query': 'Ddddddd<Read> -> u64',
        'others': [
            { 'path': 'generics_impl::Ddddddd', 'name': 'ggggggg' },
        ],
    },
    {
        'query': 'trait:Read -> u64',
        'others': [
            { 'path': 'generics_impl::Ddddddd', 'name': 'eeeeeee' },
        ],
    },
    {
        'query': 'struct:Read -> u64',
        'others': [],
    },
    {
        'query': 'bool -> u64',
        'others': [
            { 'path': 'generics_impl::Ddddddd', 'name': 'fffffff' },
        ],
    },
    {
        'query': 'Ddddddd -> u64',
        'others': [
            { 'path': 'generics_impl::Ddddddd', 'name': 'ggggggg' },
        ],
    },
    {
        'query': '-> Ddddddd',
        'others': [
            { 'path': 'generics_impl::Ddddddd', 'name': 'hhhhhhh' },
        ],
    },
];

File: tests/rustdoc-js/generics-impl.rs
use std::io::{self, Read};

pub struct Aaaaaaa;

impl Aaaaaaa {
    pub fn bbbbbbb(self) -> u32 {
        1
    }
    pub fn ccccccc(&self) -> bool {
        true
    }
}

impl Read for Aaaaaaa {
    fn read(&mut self, out: &mut [u8]) -> io::Result<usize> {
        Ok(out.len())
    }
}

pub struct Ddddddd<T>(T);

impl<T: Read> Ddddddd<T> {
    pub fn eeeeeee(_: T) -> u64 {
        1
    }
    pub fn fffffff(_: bool) -> u64 {
        1
    }
    pub fn ggggggg(self) -> u64 {
        1
    }
    pub fn hhhhhhh() -> Self
    where
        T: Default,
    {
        Ddddddd(T::default())
    }
}
